test/parallel_api/iterator/input_data_sweep.pass.cpp

```diff
@@ -0,0 +1,330 @@
+                         /*__check_write=*/false, /*__usable_as_perm_map=*/true, /*__usable_as_perm_src=*/true,
+                         /*__is_reversible=*/true>(device_policy1, my_counting, my_counting + n, counting, copy_out,
+                                                   my_counting, copy_out, counting, trash,
+                                                   std::string("counting_iterator<") + type_text + std::string(">"));
+        }
+
+        //host iterator
+        std::vector<T> host_iter(n);
+        wrap_recurse<__recurse, 0, /*__read =*/true, /*__reset_read=*/true, /*__write=*/true,
+                     /*__check_write=*/true, /*__usable_as_perm_map=*/true, /*__usable_as_perm_src=*/false,
+                     /*__is_reversible=*/true>(device_policy2, host_iter.begin(), host_iter.end(), counting, copy_out,
+                                               host_iter.begin(), copy_out, counting, trash,
+                                               std::string("host_iterator<") + type_text + std::string(">"));
+
+        //sycl iterator
+        sycl::buffer<T> buf(n);
+        //test all modes / wrappers
+        wrap_recurse<__recurse, 0, /*__read =*/true, /*__reset_read=*/true, /*__write=*/true,
+                     /*__check_write=*/true, /*__usable_as_perm_map=*/true, /*__usable_as_perm_src=*/false,
+                     /*__is_reversible=*/false>(device_policy3, oneapi::dpl::begin(buf), oneapi::dpl::end(buf),
+                                                counting, copy_out, oneapi::dpl::begin(buf), copy_out, counting, trash,
+                                                std::string("sycl_iterator<") + type_text + std::string(">"));
+
+        //usm_shared
+        auto usm_shared = sycl::malloc_shared<T>(n, device_policy.queue());
+        //test all modes / wrappers
+        wrap_recurse<__recurse, 0>(device_policy4, usm_shared, usm_shared + n, counting, copy_out, usm_shared, copy_out,
+                                   counting, trash, std::string("usm_shared<") + type_text + std::string(">"));
+
+        //usm_device
+        auto usm_device = sycl::malloc_device<T>(n, device_policy.queue());
+        //test all modes / wrappers
+        wrap_recurse<__recurse, 0>(device_policy5, usm_device, usm_device + n, counting, copy_out, usm_device, copy_out,
+                                   counting, trash, std::string("usm_device<") + type_text + std::string(">"));
+    }
+    else
+    {
+        TestUtils::unsupported_types_notifier(device_policy.queue().get_device());
+    }
+}
+
+#endif //TEST_DPCPP_BACKEND_PRESENT
+
+int
+main()
+{
+#if TEST_DPCPP_BACKEND_PRESENT
+
+    constexpr size_t n = 10;
+
+    auto q = sycl::queue{sycl::default_selector_v};
+
+    //baseline with no wrapping
+    test<float, 0>(TestUtils::make_device_policy<class Kernel1>(q), -666.0f, n, "float");
+    test<double, 0>(TestUtils::make_device_policy<class Kernel2>(q), -666.0, n, "double");
```

Comment:
What if `double` type isn't supported on device?

Reply:
I check for this within `test`, using `TestUtils::has_types_support`, and skip types which are unsupported.

Reply:
ok